### PR TITLE
JERMaker for UL AK4/AK8 jet collections and adding MCCorr201Xv9 chain.

### PR DESCRIPTION
**Pull request by TomasKello**
**JERMaker** primary serves as a tool to apply JER smearing on top of nominal jet pt collection. 
This is an updated version which applies smearing both on **AK4 (CleanJet, SubJet) and AK8 (FatJet) collections**.
In case of AK8 collection, it is also possible to apply Jet mass resolution (JMR) and scale corrections (JMS) on both "mass" and groomed "msoftdrop" mass branches. Please note that for UL samples all JMR,JMS and Puppi corrections are equal to unity (but not JER smearing).

**Optional:** JERMaker can be used to store variation branches for both pt and mass (groomed and non-groomed) for all sources of correction uncertainties (JER, JMR, JMS). It is possible to create correlated variations in 6 bins of jet pt and eta OR create one merged general variation branch. 

Note 1: JER uncertainties are turned OFF for this module (but fully functional) as we are planning to run all variations in newly developed all-in-one tool.    

Note 2: commit also adds MCCorr201Xv9 chains

### CODE REVIEW

**Comment by SiewYan** [comment]
@TomasKello , i ran the chains:

```
--------------------------------------------------------------------------------------------
   _  __                 ___           __    ___                           _          
  / |/ /__ ____  ___    / _ \___  ___ / /_  / _ \_______  _______ ___ ___ (_)__  ___ _
 /    / _ `/ _ \/ _ \  / ___/ _ \(_-</ __/ / ___/ __/ _ \/ __/ -_|_-<(_-</ / _ \/ _ `/
/_/|_/\_,_/_//_/\___/ /_/   \___/___/\__/ /_/  /_/  \___/\__/\__/___/___/_/_//_/\_, / 
                                                                               /___/  
--------------------------------------------------------------------------------------------

 Productions :  ['Summer20UL18_106x_nAODv9_Full2018v9']
 Steps       :  ['MCCorr2018v9']
_LocalSite  =  cern
_TargetSite =  None
----------- Running on production: Summer20UL18_106x_nAODv9_Full2018v9
---------------- for Step :  MCCorr2018v9
INFO: Using Interactive command
Computing baseW for VBS_SSWW_cHW_INT_test
{'HiggsMass': 0.0, 'ProdMode': 'unknown', 'DecayMode': 'unknown', 'Energy': '13TeV', 'Sample': 'VBS_SSWW_cHW_INT_test', 'FinalStateBR': 1.0, 'FinalState': 'unknown', 'xs': 0.0, 'HiggsBR': 0.0, 'ProdXS': 0.0}
{'HiggsMass': 0.0, 'ProdMode': 'unknown', 'DecayMode': 'unknown', 'Energy': '13TeV', 'Sample': 'VBS_SSWW_cHW_INT_test', 'FinalStateBR': 1.0, 'FinalState': 'unknown', 'xs': 0.0, 'HiggsBR': 0.0, 'ProdXS': 0.0}
VBS_SSWW_cHW_INT_test 0.0005061 1.000
baseW: xs,N -> W 0.0005061 7100.0 7.12816901408e-05
7.12816901408e-05
[WARNING] JMR values ignored. (To be used with fat jets only.)
[WARNING] JMS values ignored. (To be used with fat jets only.)
JERMaker: 
    JetType = AK4PFchs
    JetColl = CleanJet
    JER input = Summer19UL18_JRV2_MC_PtResolution_AK4PFchs.txt
    JER SF input = Summer19UL18_JRV2_MC_SF_AK4PFchs.txt
    JER uncty = False
    splitJER = False
    doGroomed = False
[WARNING] Warning PUPPI corrections do not exist for UL samples.
JERMaker: 
    JetType = AK8PFPuppi
    JetColl = FatJet
    JER input = Summer19UL18_JRV2_MC_PtResolution_AK8PFPuppi.txt
    JER SF input = Summer19UL18_JRV2_MC_SF_AK8PFPuppi.txt
    JER uncty = False
    splitJER = False
    doGroomed = True
Load Library 'CondFormatsBTauObjects'
Load Library 'CondToolsBTau'
Loading macros from /afs/cern.ch/work/s/shoh/analysis/WH-Study/my-latino/producer/CMSSW_10_6_27/src/LatinoAnalysis/NanoGardener/python/modules/RoccoR.cc
Loaded
scale factors from /afs/cern.ch/work/s/shoh/analysis/WH-Study/my-latino/producer/CMSSW_10_6_27/src/LatinoAnalysis/NanoGardener/python/data/RoccoR2018UL.txt
TrigMaker: CMSSW = Full2018v9, isData = False, keepRunPeriod = False
LeptonSFMaker: making scale factors for analysis of Full2018v9
Will write selected trees to .
Loading jet energy resolutions (JER) from file '/tmp/shoh/tmpMKjO2c/Summer19UL18_JRV2_MC_PtResolution_AK4PFchs.txt'
Loading JER scale factors and uncertainties from file '/tmp/shoh/tmpMKjO2c/Summer19UL18_JRV2_MC_SF_AK4PFchs.txt'
Loading jet energy resolutions (JER) from file '/tmp/shoh/tmpeyHOy6/Summer19UL18_JRV2_MC_PtResolution_AK8PFPuppi.txt'
Loading JER scale factors and uncertainties from file '/tmp/shoh/tmpeyHOy6/Summer19UL18_JRV2_MC_SF_AK8PFPuppi.txt'
Pre-select 13026 entries out of 13026 (100.00%)
[WARNING] Original jet collection mass branch will be backed by Jet_mass
Computing PU profile for this file
/afs/cern.ch/work/s/shoh/analysis/WH-Study/my-latino/producer/CMSSW_10_6_27/python/LatinoAnalysis/NanoGardener/modules/Grafter.py:37: RuntimeWarning: creating converter for unknown type "const char*[]"
  self.out.fillBranch(var, self.variablesDecoded[var].EvalInstance())
Traceback (most recent call last):
  File "/afs/cern.ch/work/s/shoh/analysis/WH-Study/my-latino/producer/CMSSW_10_6_27/src/jobs//NanoGardening__Summer20UL18_106x_nAODv9_Full2018v9__MCCorr2018v9/VBS_SSWW_cHW_INT_test/NanoGardening__Summer20UL18_106x_nAODv9_Full2018v9__MCCorr2018v9__VBS_SSWW_cHW_INT_test__part0____MCl1loose2018v9.py", line 124, in <module>
    p.run() 
  File "/afs/cern.ch/work/s/shoh/analysis/WH-Study/my-latino/producer/CMSSW_10_6_27/python/PhysicsTools/NanoAODTools/postprocessing/framework/postprocessor.py", line 234, in run
    eventRange=eventRange, maxEvents=self.maxEntries
  File "/afs/cern.ch/work/s/shoh/analysis/WH-Study/my-latino/producer/CMSSW_10_6_27/python/PhysicsTools/NanoAODTools/postprocessing/framework/eventloop.py", line 82, in eventLoop
    ret = m.analyze(e)
  File "/afs/cern.ch/work/s/shoh/analysis/WH-Study/my-latino/producer/CMSSW_10_6_27/python/LatinoAnalysis/NanoGardener/modules/BoostedWtagSF.py", line 89, in analyze
    fj_softdrop_mass = getattr(fj, 'msoftdrop' + self._input_branch_prefix)
  File "/afs/cern.ch/work/s/shoh/analysis/WH-Study/my-latino/producer/CMSSW_10_6_27/python/PhysicsTools/NanoAODTools/postprocessing/framework/datamodel.py", line 69, in __getattr__
    val = getattr(self._event, self._prefix + name)
  File "/afs/cern.ch/work/s/shoh/analysis/WH-Study/my-latino/producer/CMSSW_10_6_27/python/PhysicsTools/NanoAODTools/postprocessing/framework/datamodel.py", line 18, in __getattr__
    return self._tree.readBranch(name)
  File "/afs/cern.ch/work/s/shoh/analysis/WH-Study/my-latino/producer/CMSSW_10_6_27/python/PhysicsTools/NanoAODTools/postprocessing/framework/treeReaderArrayTools.py", line 80, in readBranch
    raise RuntimeError("Unknown branch %s" % branchName)
RuntimeError: Unknown branch FatJet_msoftdrop_nom
```

Maybe those messages below:

```
Loading jet energy resolutions (JER) from file '/tmp/shoh/tmpMKjO2c/Summer19UL18_JRV2_MC_PtResolution_AK4PFchs.txt'
Loading JER scale factors and uncertainties from file '/tmp/shoh/tmpMKjO2c/Summer19UL18_JRV2_MC_SF_AK4PFchs.txt'
Loading jet energy resolutions (JER) from file '/tmp/shoh/tmpeyHOy6/Summer19UL18_JRV2_MC_PtResolution_AK8PFPuppi.txt'
Loading JER scale factors and uncertainties from file '/tmp/shoh/tmpeyHOy6/Summer19UL18_JRV2_MC_SF_AK8PFPuppi.txt'
```

can be omitted follow by the  ```JER uncty = False```.

**Comment by TomasKello** [comment]
Hi @SiewYan this error you can see is coming from BoostedWtagSF module which indeed operates with `FatJet_msoftdrop_nom` which now does not exist as I mentioned [here](https://mattermost.web.cern.ch/latinos/pl/atfzbgfmkjgrmgw5xpkm7t9cbo). As there is only nominal pT driven JER smearing applied on softdrop mass, there is no need for `_nom` extension. This way we can avoid running JERMaker over Data (only to create branch with this extension) and only run module on MC. As a consequence it is needed to rename this variable in BoostedWtagSF module and possibly elsewhere. We can add this into this commit indeed. 

Not sure what you mean by omitting "Loading ..." messages. They are harmless.

File: NanoGardener/python/framework/Steps_cfg.py [hunk 1]
```diff
@@ -3252,7 +3276,24 @@ def prepare_CombJJLNu_syst(basename, selection):
...
+                  'declare'    : 'JERMakerMCUL = lambda : JERMaker(RPLME_YEAR, RPLME_CMSSW,"",jetType="AK4PFchs",jetColl="CleanJet")',
+                  'module'     : 'JERMakerMCUL()',
+                 },
+
+    'FatJERsMCUL': {
+                  'isChain'    : False ,
+                  'do4MC'      : True  ,
+                  'do4Data'    : False  ,
+                  'import'     : 'LatinoAnalysis.NanoGardener.modules.JERMaker' ,
+                  'declare'    : 'FatJERMakerMCUL = lambda : JERMaker(RPLME_YEAR, RPLME_CMSSW,"",jetType="AK8PFPuppi",jetColl="FatJet",doGroomed=True)',
```
[the remark this answers is not in the document]
Comment (SiewYan) [reply NanoGardener/python/framework/Steps_cfg.py]: resolved

File: NanoGardener/python/framework/Steps_cfg.py [hunk 2]
```diff
@@ -3252,7 +3276,24 @@ def prepare_CombJJLNu_syst(basename, selection):
...
                   'module'     : 'JERMakerMC18()',
                  },
-    #jerTag for 2018 is missing on purpose, 2017 JERs are used instead (for a moment)
+
+    'JERsMCUL': {
+                  'isChain'    : False ,
+                  'do4MC'      : True  ,
+                  'do4Data'    : False  ,
+                  'import'     : 'LatinoAnalysis.NanoGardener.modules.JERMaker' ,
+                  'declare'    : 'JERMakerMCUL = lambda : JERMaker(RPLME_YEAR, RPLME_CMSSW,"",jetType="AK4PFchs",jetColl="CleanJet")',
```
[the remark this answers is not in the document]
Comment (SiewYan) [reply NanoGardener/python/framework/Steps_cfg.py]: resolved

**Comment by TomasKello** [comment]
Thanks a lot! I am suggesting to merge this PR now. Alternatively we can fix/remove WtagSF module in another PR.

**Pull request merged**